src/Analysis/Engine/Impl/Values/ModuleInfo.cs

```diff
@@ -172,22 +161,27 @@ internal class ModuleInfo : AnalysisValue, IReferenceableContainer, IModuleInfo
             return null;
         }
 
-        public void AddModuleReference(ModuleReference moduleRef) {
-            if (moduleRef == null) {
-                Debug.Fail("moduleRef should never be null");
-                throw new ArgumentNullException(nameof(moduleRef));
+        public bool TryGetModuleReference(string moduleName, out ModuleReference moduleReference) {
+            lock (_projectEntry) {
+                return _referencedModules.TryGetValue(moduleName, out moduleReference);
             }
-            _referencedModules.Add(moduleRef);
-            moduleRef.AddReference(this);
         }
 
-        public void RemoveModuleReference(ModuleReference moduleRef) {
-            if (_referencedModules.Remove(moduleRef)) {
-                moduleRef.RemoveReference(this);
+        public void AddModuleReference(string moduleName, ModuleReference moduleReference) {
+            Check.ArgumentNotNull(nameof(moduleName), moduleName);
+            Check.ArgumentNotNull(nameof(moduleReference), moduleReference);
+
+            _referencedModules[moduleName] = moduleReference;
+            moduleReference.AddReference(this);
+        }
+
+        public void ClearReferencedModules() {
+            foreach (var moduleReference in _referencedModules.Values.Distinct()) {
```

Comment:
Not sure if Values can be null (when module is missing) - perhaps `MaybeEnumerate`?

Reply:
`_referencedModules` is a dictionary, it is always initialized and it is readonly.